includes/enrolment/class-sensei-enrolment-provider-state-snapshot.php

```diff
@@ -0,0 +1,114 @@
+<?php
+/**
+ * File containing the class Sensei_Enrolment_Provider_State_Snapshot.
+ *
+ * @package sensei
+ */
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit;
+}
+
+/**
+ * A snapshot of the state of all enrolment providers.
+ */
+class Sensei_Enrolment_Provider_State_Snapshot implements JsonSerializable {
+
+	/**
+	 * Timestamp of the snapshot.
+	 *
+	 * @var int
+	 */
+	private $timestamp;
+
+	/**
+	 * An array of the status of all enrolment providers.
+	 *
+	 * [
+	 *      'provider_id' => [ 'enrolment_status' => true|false ]
+	 * ]
+	 *
+	 * @var array
+	 */
+	private $providers_status;
+
+	/**
+	 * Class constructor.
+	 *
+	 * @param array   $providers_status The providers' status.
+	 * @param integer $timestamp        The timestamp of the snapshot.
+	 */
+	private function __construct( $providers_status, $timestamp = null ) {
+		$this->timestamp        = null === $timestamp ? time() : $timestamp;
+		$this->providers_status = $providers_status;
+	}
+
+	/**
+	 * Restore a snapshot from a JSON string.
+	 *
+	 * @param array $data Serialized state of object.
+	 *
+	 * @return self|false
+	 */
+	public static function from_serialized_array( $data ) {
+
+		if ( empty( $data ) ) {
+			return false;
+		}
+
+		return new self( $data['p'], $data['t'] );
+	}
+
+	/**
+	 * Create an empty snapshot.
+	 *
+	 * @param array $providers_status The providers' status.
+	 *
+	 * @return self
+	 */
+	public static function create( $providers_status = [] ) {
+		return new self( $providers_status );
+	}
+
+	/**
+	 * Create a duplicate of a snapshot.
+	 *
+	 * @param Sensei_Enrolment_Provider_State_Snapshot $other The snapshot to copy.
+	 *
+	 * @return self
+	 */
+	public static function duplicate( Sensei_Enrolment_Provider_State_Snapshot $other ) {
+		return new self( $other->providers_status );
+	}
+
+	/**
+	 * Return object that can be serialized by `json_encode()`.
+	 *
+	 * @return array
+	 */
+	public function jsonSerialize() {
+		return [
+			't' => $this->timestamp,
+			'p' => $this->providers_status,
+		];
+	}
+
+	/**
+	 * Update the status of a provider.
+	 *
+	 * @param string $provider_id The provider to update the status for.
+	 * @param bool   $is_enrolled The enrolment status.
+	 *
+	 * @return bool True if the value was actually updated. False otherwise.
+	 */
+	public function update_status( $provider_id, $is_enrolled ) {
+
+		if ( isset( $this->providers_status[ $provider_id ]['enrolment_status'] ) && $is_enrolled === $this->providers_status[ $provider_id ]['enrolment_status'] ) {
+			return false;
+		}
+
+		$this->providers_status[ $provider_id ]['enrolment_status'] = $is_enrolled;
```

Comment:
Do we need the second dimension of this array? For something that will be stored in the database a lot, `enrolment_status` might be unnecessary. If we decide to keep `enrolment_status`, we should initialize the first dimension as an array.

Reply:
I have renamed `enrolment_status` to `es` to save space. For the second dimension I think that it's needed in case that we want to add to the history something else other than the enrolment status.

> If we decide to keep enrolment_status, we should initialize the first dimension as an array.

Can you explain this a bit?

Reply:
The `$this->providers_status[ $provider_id ]` dimension is never initialized as an array before being used as an array (unless I'm missing it).

Reply:
Ah! It is concealed a bit in the default argument of the `Sensei_Enrolment_Provider_State_Snapshot::create` method. I just added it a default argument in the constructor too in a449a3fc52b118e37f8a83d89ec27101592c6c65. (I can remove the commit if you don't think that it's needed).

Reply:
Fixed in 83976eb1f23aa8e6570736859d133232f453c32e